release-22.1: sql: correctly handle COLLATEd columns in COPY

otan opened the pull request:
Backport 2/2 commits from #95894.

/cc @cockroachdb/release

---

See individual commits for details.

Resolves #95887

Release justification: fixes a critical bug with COPY

otan committed:
tree: rename ParseTimeContext to ParseContext

A mechanical rename PR. Anything named after ParseTimeContext is now
just ParseContext.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96039)
<!-- Reviewable:end -->

otan committed:
sql: fix collated strings during COPY

Release note (bug fix): Fix a bug where COPYing into a column with
collated strings would result in an error similar to
`internal error: unknown type collatedstring`.
